avalanche/training/templates/NEW_base_sgd.py

```diff
@@ -0,0 +1,519 @@
+
+        self.loss = None
+        """ Loss of the current mini-batch. """
+
+        self._stop_training = False
+
+    def train(self,
+              experiences: Union[CLExperience,
+                                 ExpSequence],
+              eval_streams: Optional[Sequence[Union[CLExperience,
+                                                    ExpSequence]]] = None,
+              **kwargs):
+
+        super().train(experiences, eval_streams, **kwargs)
+        return self.evaluator.get_last_metrics()
+
+    @torch.no_grad()
+    def eval(self, exp_list: Union[CLExperience, CLStream], **kwargs):
+        """
+        Evaluate the current model on a series of experiences and
+        returns the last recorded value for each metric.
+
+        :param exp_list: CL experience information.
+        :param kwargs: custom arguments.
+
+        :return: dictionary containing last recorded value for
+            each metric name
+        """
+        super().eval(exp_list, **kwargs)
+        return self.evaluator.get_last_metrics()
+
+    def _train_exp(
+        self, experience: CLExperience, eval_streams, **kwargs
+    ):
+        # Should be implemented in Observation Type
+        raise NotImplementedError()
+
+    def _eval_exp(self, **kwargs):
+        self.eval_epoch(**kwargs)
+
+    def make_optimizer(self, **kwargs):
+        """Optimizer initialization."""
+        # Should be implemented in Observation Type
+        raise NotImplementedError()
+
+    def criterion(self):
+        """Compute loss function."""
+        raise NotImplementedError()
+
+    def forward(self):
+        """Compute the model's output given the current mini-batch."""
+        raise NotImplementedError()
+
+    def model_adaptation(self, model=None):
+        """Adapts the model to the current experience."""
+        raise NotImplementedError()
+
+    def stop_training(self):
+        """Signals to stop training at the next iteration."""
+        self._stop_training = True
+
+    def training_epoch(self, **kwargs):
+        # Should be implemented in Update Type
+        raise NotADirectoryError()
+
+    def backward(self):
+        """Run the backward pass."""
+        self.loss.backward()
+
+    def optimizer_step(self):
+        """Execute the optimizer step (weights update)."""
+        self.optimizer.step()
+
+    def eval_epoch(self, **kwargs):
+        """Evaluation loop over the current `self.dataloader`."""
+        for self.mbatch in self.dataloader:
+            self._unpack_minibatch()
+            self._before_eval_iteration(**kwargs)
+
+            self._before_eval_forward(**kwargs)
+            self.mb_output = self.forward()
+            self._after_eval_forward(**kwargs)
+            self.loss = self.criterion()
+
+            self._after_eval_iteration(**kwargs)
+
+    # ==================================================================> NEW
+
+    def maybe_adapt_model_and_make_optimizer(self):
```

Comment:
why do we need this method? strategies can override `_before_training_exp` for the same behavior

Reply:
if you need a call that can be implemented by `observation_type` I would use a more general name, something like `prepare_training_exp` or similar.